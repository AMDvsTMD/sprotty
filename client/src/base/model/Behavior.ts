///<reference path="GModel.ts"/>
import {Point} from "../../utils/Geometry"
import {GModelElement} from "./GModel"

export interface Behavior {
}

export interface Moveable extends Behavior, Point {
    x: number
    y: number
}

export function isMoveable(element: GModelElement | Moveable): element is Moveable {
    return 'x' in element && 'y' in element
}

export interface Selectable extends Behavior {
    selected: boolean
}

export function isSelectable(element: GModelElement | Selectable): element is Selectable {
    return 'selected' in element
}

