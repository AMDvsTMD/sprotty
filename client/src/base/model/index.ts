export * from './behavior';
export * from './gmodel';
export * from './gmodel-schema';
export * from './gmodel-source';
export * from './gmodel-factory';
