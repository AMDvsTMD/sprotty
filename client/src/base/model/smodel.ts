
export interface SModelElementSchema {
    type: string
    id: string
}

/**
 * Base class for all elements of the diagram model.
 * Each model element must have a unique ID and a type that is used to look up its view.
 */
export class SModelElement implements SModelElementSchema {
    type: string
    id: string

    get root(): SModelRoot {
        let current: SModelElement | undefined = this
        while (current) {
            if (current instanceof SModelRoot)
                return current
            else if (current instanceof SChildElement)
                current = current.parent
            else
                current = undefined
        }
        throw new Error("Element has no root")
    }

    get index(): SModelIndex {
        return this.root.index
    }

    hasFeature(feature: symbol): boolean {
        return false
    }
}

export interface SParentElementSchema extends SModelElementSchema {
    children?: SModelElementSchema[]
}

export class SParentElement extends SModelElement implements SParentElementSchema {
    children: SChildElement[] = []

    add(child: SChildElement, i?: number) {
        if (i === undefined) {
            this.children.push(child)
        } else {
            if (i < 0 || i > this.children.length) {
                throw "Child index out of bounds " + i + " (0.." + this.children.length + ")"
            }
            this.children.splice(i, 0, child)
        }
        child.parent = this
        this.index.add(child)
    }

    remove(child: SChildElement) {
        const i = this.children.indexOf(child)
        if (i < 0) {
            throw "No such child " + child
        }
        this.children.splice(i, 1)
        this.index.remove(child)
    }

    move(child: SChildElement, newIndex: number) {
        const i = this.children.indexOf(child)
        if (i == -1) {
            throw "No such child " + child
        } else {
            if (newIndex < 0 || newIndex > this.children.length - 1) {
                throw "Child index out of bounds " + i + " (0.." + this.children.length + ")"
            }
            this.children.splice(i, 1)
            this.children.splice(newIndex, 0, child)
        }
    }
}

export class SChildElement extends SParentElement {
    parent: SParentElement
}

export interface SModelRootSchema extends SParentElementSchema {
}

/**
 * Base class for the root elements of the diagram model tree.
 */
export class SModelRoot extends SParentElement implements SModelRootSchema {
    private _index: SModelIndex

    get index(): SModelIndex {
        if (!this._index) {
            this._index = new SModelIndex
            this._index.add(this)
        }
        return this._index
    }
}

/**
 * Used to speed up model element lookup by id.
 */
export class SModelIndex {

    private id2element: Map<string, SModelElement> = new Map

    add(element: SModelElement): void {
        this.id2element.set(element.id, element)
    }

    remove(element: SModelElement): void {
        this.id2element.delete(element.id)
    }

    contains(element: SModelElement): boolean {
        return this.id2element.get(element.id) !== undefined
    }

    removeById(elementId: string): void {
        this.id2element.delete(elementId)
    }

    getById(id: string): SModelElement | undefined {
        return this.id2element.get(id)
    }

    all(): SModelElement[] {
        const all: SModelElement[] = []
        this.id2element.forEach(
            element => all.push(element)
        )
        return all
    }
}

export function getBasicType(schema: SModelElementSchema): string {
    if (!schema.type)
        return ''
    let colonIndex = schema.type.indexOf(':')
    if (colonIndex >= 0)
        return schema.type.substring(0, colonIndex)
    else
        return schema.type
}

export function getParent<T>(element: SModelElement | T, predicate: (e: SModelElement) => boolean): (SModelElement & T) | undefined {
    if (predicate.call(undefined, element))
        return element as (SModelElement & T)
    else if (element instanceof SChildElement) {
        const parent = element.parent
        if (parent)
            return getParent<T>(parent, predicate)
    }
    return undefined
}
