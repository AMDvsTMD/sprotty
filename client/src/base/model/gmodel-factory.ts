import {GGraph, GGraphSchema} from "../../graph/model"
import {GModelRootSchema} from "./gmodel-schema"
import {GModelRoot} from "./gmodel"

export namespace GModelFactory {

    export function createModel(schema: GModelRootSchema): GModelRoot {
        if (GGraphSchema.isGGraphSchema(schema))
            return new GGraph(schema)
        else
            return new GModelRoot(schema)
    }

}
