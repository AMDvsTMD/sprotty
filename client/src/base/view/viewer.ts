import {h, init} from "snabbdom"
import {VNode} from "snabbdom/vnode"
import {Module} from "snabbdom/modules/module"
import {classModule} from "snabbdom/modules/class"
import {propsModule} from "snabbdom/modules/props"
import {attributesModule} from "snabbdom/modules/attributes"
import {styleModule} from "snabbdom/modules/style"
import {eventListenersModule} from "snabbdom/modules/eventlisteners"
import {CommandStackCallback, Action} from "../intent"
import {SModelRoot, SModelElement, SParentElement} from "../model"
import {EventSource} from "../../utils"
import {AddRemoveAnimationDecorator, VNodeDecorator} from "./vnode-decorators"
import {RenderingContext, ViewRegistry} from "./views"
import {KeyTool} from "./key-tool"
import {MouseTool} from "./mouse-tool"

/**
 * The component that turns the model into an SVG DOM.
 * Uses a VDOM based on snabbdom.js for performance.
 */
export class Viewer extends EventSource<ViewerCallback> implements CommandStackCallback, VNodeDecorator {

    viewRegistry = new ViewRegistry()
    patcher: Patcher
    lastVDOM: undefined
    decorators: VNodeDecorator[] = []

    constructor(private baseDiv: string) {
        super()
        this.patcher = this.createPatcher()
        this.decorators = this.createDecorators()
    }

    createDecorators(): VNodeDecorator[] {
        return [new AddRemoveAnimationDecorator(), new KeyTool(this), new MouseTool(this)]
    }

    createModules(): Module[] {
        return [
            classModule,
            propsModule,
            attributesModule,
            styleModule,
            eventListenersModule
        ]
    }

    createPatcher() {
        return init(this.createModules())
    }

    createRenderingContext(model: SModelRoot): RenderingContext {
        return {
            viewer: this,
        }
    }

    decorate(vnode: VNode, element: SModelElement): VNode {
        return this.decorators.reduce(
            (vnode: VNode, decorator: VNodeDecorator) => decorator.decorate(vnode, element),
            vnode)
    }

    renderElement(element: SModelElement, context: RenderingContext): VNode {
        const vNode = this.viewRegistry.get(element.type, element).render(element, context)
        return this.decorate(vNode, element)
    }

    renderChildren(element: SParentElement, context: RenderingContext): VNode[] {
        return element.children.map((element) => context.viewer.renderElement(element, context))
    }

    update(model: SModelRoot): void {
        const context = this.createRenderingContext(model)
        const newVDOM = h('div', {
            attrs: {
                id: this.baseDiv,
                class: this.baseDiv
            }
        }, [
            this.renderElement(model, context)
        ])
        if (this.lastVDOM) {
            this.lastVDOM = this.patcher.call(this, this.lastVDOM, newVDOM)
        } else {
            const placeholder = document.getElementById(this.baseDiv)
            this.lastVDOM = this.patcher.call(this, placeholder, newVDOM)
        }
    }

    fireAction(action: Action) {
        this.callbacks.forEach(callback => callback.execute([action]))
    }
}

export type Patcher = (oldRoot: VNode | Element, newRoot: VNode) => VNode

export interface ViewerCallback {
    execute(actions: Action[]): void
}

