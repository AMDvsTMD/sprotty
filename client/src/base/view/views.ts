import "reflect-metadata"
import { injectable } from "inversify"
import {VNode} from "snabbdom/vnode"
import {h} from "snabbdom"
import {SModelElement, SModelRoot, SModel} from "../model"
import {ProviderRegistry} from "../../utils"
import {Viewer} from "./viewer"

/**
 * Base interface for the components that turn GModelElements into virtual DOM elements.
 */
export interface View {
    render(model: SModelElement, context: RenderingContext): VNode
}

/**
 * Bundles additional data that is passed to views for VNode creation.
 */
export interface RenderingContext {
    viewer: Viewer
}

/**
 * Allows to look up the View for a given SModelElement based on its type.
 */
@injectable()
export class ViewRegistry extends ProviderRegistry<View, SModelElement> {
    constructor() {
        super()
        this.registerDefaults()
    }

    protected registerDefaults() {
        this.register(SModel.EMPTY_ROOT.type, EmptyView)
    }

    missing(key: string, element: SModelElement) {
        return new MissingView()
    }
}

export class EmptyView implements View {
    render(model: SModelRoot, context: RenderingContext): VNode {
        return h('g', {
            key: model.id,
            attrs: {
                id: model.id
            }
        });
    }
}

export class MissingView implements View {
    render(model: SModelElement, context: RenderingContext): VNode {
        const x = (model as any).x || 0
        const y = (model as any).y || 0
        return h('text', {
            class: {
                missing: true
            },
            attrs: {
                'text-anchor': 'middle',
                x: x,
                y: y,
            }
        }, '?' + model.id + '?')
    }
}