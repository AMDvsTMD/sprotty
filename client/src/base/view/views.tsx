import * as snabbdom from "snabbdom-jsx"
import { injectable } from "inversify"
import { VNode } from "snabbdom/vnode"
import { SModelElement, SModelRoot } from "../model/smodel"
import { EMPTY_ROOT } from "../model/smodel-factory"
import { ProviderRegistry } from "../../utils/registry"
import { Viewer } from "./viewer"

const JSX = {createElement: snabbdom.svg}

/**
 * Base interface for the components that turn GModelElements into virtual DOM elements.
 */
export interface View {
    render(model: SModelElement, context: RenderingContext): VNode
}

/**
 * Bundles additional data that is passed to views for VNode creation.
 */
export interface RenderingContext {
    viewer: Viewer
}

/**
 * Allows to look up the View for a given SModelElement based on its type.
 */
@injectable()
export class ViewRegistry extends ProviderRegistry<View, SModelElement> {
    constructor() {
        super()
        this.registerDefaults()
    }

    protected registerDefaults() {
        this.register(EMPTY_ROOT.type, EmptyView)
    }

    missing(key: string, element: SModelElement) {
        return new MissingView()
    }
}

export class EmptyView implements View {
    render(model: SModelRoot, context: RenderingContext): VNode {
        return <svg key={model.id} id={model.id} class-empty={true} />
    }
}

export class MissingView implements View {
    render(model: SModelElement, context: RenderingContext): VNode {
        const x = (model as any).x || 0
        const y = (model as any).y || 0
        return <text class-missing={true} id={model.id} x={x} y={y}>?{model.id}?</text>
    }
}