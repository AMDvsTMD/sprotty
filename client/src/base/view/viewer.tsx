import { inject, injectable, multiInject, optional } from "inversify"
import * as snabbdom from "snabbdom-jsx"
import { init } from "snabbdom"
import { VNode } from "snabbdom/vnode"
import { Module } from "snabbdom/modules/module"
import { propsModule } from "snabbdom/modules/props"
import { attributesModule } from "snabbdom/modules/attributes"
import { styleModule } from "snabbdom/modules/style"
import { eventListenersModule } from "snabbdom/modules/eventlisteners"
import { classModule } from "snabbdom/modules/class"
import { SModelElement, SModelRoot, SParentElement } from "../model/smodel"
import { TYPES } from "../types"
import { VNodeDecorator } from "./vnode-decorators"
import { RenderingContext, ViewRegistry } from "./views"
import { setClass } from "./vnode-utils"
import { IViewerOptions } from "./options"
import { ILogger } from "../../utils/logging"
import { isThunk } from "./thunk-view"

const JSX = {createElement: snabbdom.html}  // must be html here, as we're creating a div

export interface IViewer {
    update(model: SModelRoot): void
}

/**
 * The component that turns the model into an SVG DOM.
 * Uses a VDOM based on snabbdom.js for performance.
 */
@injectable()
export class Viewer implements VNodeDecorator, IViewer {

    @inject(TYPES.ILogger) public logger: ILogger
    @inject(TYPES.ViewRegistry) public viewRegistry: ViewRegistry
    @inject(TYPES.IViewerOptions) protected options: IViewerOptions

    protected readonly patcher: Patcher
    protected lastVDOM: VNode

    constructor(@multiInject(TYPES.VNodeDecorator)@optional() protected decorators: VNodeDecorator[]) {
        this.patcher = this.createPatcher()
    }

    protected createModules(): Module[] {
        return [
            propsModule,
            attributesModule,
            classModule,
            styleModule,
            eventListenersModule
        ]
    }

    protected createPatcher() {
        return init(this.createModules())
    }

    protected createRenderingContext(model: SModelRoot): RenderingContext {
        return {
            viewer: this,
        }
    }

    decorate(vnode: VNode, element: SModelElement): VNode {
        if(isThunk(vnode))
            return vnode
        this.decorators = this.decorators
        return this.decorators.reduce(
            (vnode: VNode, decorator: VNodeDecorator) => decorator.decorate(vnode, element),
            vnode)
    }

    postUpdate() {
        this.decorators.forEach(decorator => decorator.postUpdate())
    }

    renderElement(element: SModelElement, context: RenderingContext): VNode {
        const vNode = this.viewRegistry.get(element.type, element).render(element, context)
        return this.decorate(vNode, element)
    }

    renderChildren(element: SParentElement, context: RenderingContext): VNode[] {
        return element.children.map((element) => context.viewer.renderElement(element, context))
    }

    update(model: SModelRoot): void {
        this.logger.log(this, 'rendering', JSON.stringify((model as any).boundsInPage))
        const context = this.createRenderingContext(model)
        const newVDOM = <div id={this.options.baseDiv}>
                {this.renderElement(model, context) as VNode}
            </div>
        setClass(newVDOM, this.options.baseDiv, true)
        if (this.lastVDOM) {
            this.lastVDOM = this.patcher.call(this, this.lastVDOM, newVDOM)
        } else {
            const placeholder = document.getElementById(this.options.baseDiv)
            this.lastVDOM = this.patcher.call(this, placeholder, newVDOM)
        }
        this.postUpdate()
    }

}

export type Patcher = (oldRoot: VNode | Element, newRoot: VNode) => VNode

export type IViewerProvider = () => Promise<Viewer>
