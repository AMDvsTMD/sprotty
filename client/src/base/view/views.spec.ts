import {SModel} from "../model/smodel"
import {EmptyView, MissingView} from "./views"
import {Viewer} from "./viewer"
import {SNode} from "../../graph/model/sgraph"
const toHTML = require('snabbdom-to-html')

describe('base views', () => {

    const context = {viewer: new Viewer()}

    it('empty view', () => {
        const emptyView = new EmptyView()
        const vnode = emptyView.render(SModel.EMPTY_ROOT, context)
        const html = toHTML(vnode)
        expect(html).toMatchSnapshot()
    })

    const missingView = new MissingView

    it('missing view', () => {
        const vnode = missingView.render(SModel.EMPTY_ROOT, context)
        expect(toHTML(vnode)).toMatchSnapshot()
        const model = new SNode()
        model.x = 42
        model.y = 41
        model.id = 'foo'
        model.type = 'type'
        const vnode1 = missingView.render(model, context)
        expect(toHTML(vnode1)).toMatchSnapshot()
    })
})
