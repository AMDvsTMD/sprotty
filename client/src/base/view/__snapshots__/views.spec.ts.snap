// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base views empty view 1`] = `"<g id=\\"EMPTY\\"></g>"`;

exports[`base views missing view 1`] = `"<text class=\\"missing\\" text-anchor=\\"middle\\" x=\\"0\\" y=\\"0\\">?EMPTY?</text>"`;

exports[`base views missing view 2`] = `"<text class=\\"missing\\" text-anchor=\\"middle\\" x=\\"42\\" y=\\"41\\">?foo?</text>"`;
